src/vm/gcenv.ee.cpp

```diff
@@ -932,40 +934,48 @@ void GCToEEInterface::StompWriteBarrier(WriteBarrierParameters* args)
         
         g_lowest_address = args->lowest_address;
         VolatileStore(&g_highest_address, args->highest_address);
-        ::StompWriteBarrierResize(true, false);
+        flushXrestart |= ::StompWriteBarrierResize(true, false);
 
         // StompWriteBarrierResize does not necessarily bash g_ephemeral_low
         // usages, so we must do so here. This is particularly true on x86,
         // where StompWriteBarrierResize will not bash g_ephemeral_low when
         // called with the parameters (true, false), as it is above.
         g_ephemeral_low = args->ephemeral_low;
         g_ephemeral_high = args->ephemeral_high;
-        ::StompWriteBarrierEphemeral(true);
-        return;
+        flushXrestart |= ::StompWriteBarrierEphemeral(true);
+        break;
     case WriteBarrierOp::SwitchToWriteWatch:
 #ifdef FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
         assert(args->write_watch_table != nullptr);
         assert(args->is_runtime_suspended && "the runtime must be suspended here!");
         g_sw_ww_table = args->write_watch_table;
         g_sw_ww_enabled_for_gc_heap = true;
-        ::SwitchToWriteWatchBarrier(true);
+        flushXrestart |= ::SwitchToWriteWatchBarrier(true);
 #else
         assert(!"should never be called without FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP");
 #endif // FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
-        return;
+        break;
     case WriteBarrierOp::SwitchToNonWriteWatch:
 #ifdef FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
         assert(args->is_runtime_suspended && "the runtime must be suspended here!");
         g_sw_ww_table = 0;
         g_sw_ww_enabled_for_gc_heap = false;
-        ::SwitchToNonWriteWatchBarrier(true);
+        flushXrestart |= ::SwitchToNonWriteWatchBarrier(true);
 #else
         assert(!"should never be called without FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP");
 #endif // FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
-        return;
+        break;
     default:
         assert(!"unknown WriteBarrierOp enum");
     }
+    if (flushXrestart & ICACHE_FLUSH) 
+    {
+        ::FlushWriteBarrierInstructionCache();
+    }
+    if (flushXrestart & EE_RESTART) 
```

Comment:
would be nice to have an `assert(args->is_runtime_suspended)` here, to be sure we're not going to accidentally try to restart the EE when it was never suspended.

Reply:
But this method can be invoked with running runtime which may be suspended inside `StompWriteBarrierXXX`. That would be false alarm in such case.

Reply:
Sorry, I typo'd the assert, I meant `assert(!args->is_runtime_suspended)`. If I'm understanding your invariants right, the presence of the `EE_RESTART` flag implies that the runtime was **not** suspended, was suspended by a write barrier stomp, and now needs to be restarted, correct?

Reply:
That's right. I will put it here.